nixos/modules/config/mysql.nix

```diff
@@ -0,0 +1,485 @@
+              example = "password";
+              description = "The name of the column that contains a (encrypted) password string.";
+            };
+            statusColumn = mkOption {
+              type = types.nullOr types.str;
+              default = null;
+              example = "status";
+              description = ''
+                The name of the column or an SQL expression that indicates the status of
+                the user. The status is expressed by the combination of two bitfields
+                shown below:
+
+                bit 0 (0x01): if flagged, <literal>pam_mysql</literal> deems the account to be expired and
+                              returns <literal>PAM_ACCT_EXPIRED</literal>. That is, the account is supposed
+                              to no longer be available. Note this doesn't mean that <literal>pam_mysql</literal>
+                              rejects further authentication operations.
+
+                bit 1 (0x02): if flagged, <literal>pam_mysql</literal> deems the authentication token
+                              (password) to be expired and returns <literal>PAM_NEW_AUTHTOK_REQD</literal>.
+                              This ends up requiring that the user enter a new password.
+              '';
+            };
+            passwordCrypt = mkOption {
+              example = "2";
+              type = types.enum [
+                "0" "plain"
+                "1" "Y"
+                "2" "mysql"
+                "3" "md5"
+                "4" "sha1"
+                "5" "drupal7"
+                "6" "joomla15"
+                "7" "ssha"
+                "8" "sha512"
+                "9" "sha256"
+              ];
+              description = ''
+                The method to encrypt the user's password:
+
+                0 (or "plain") 	= No encryption. Passwords are stored in plaintext.
+                                  HIGHLY DISCOURAGED.
+
+                1 (or "Y")      = Use crypt(3) function.
+
+                2 (or "mysql")  = Use the MySQL PASSWORD() function. It is possible
+                                  that the encryption function used by <literal>
+                                  pam_mysql</literal> is different from that of
+                                  the MySQL server, as <literal>pam_mysql</literal>
+                                  uses the function defined in MySQL's C-client API
+                                  instead of using PASSWORD() SQL function in the query.
+
+                3 (or "md5")     = Use plain hex MD5.
+
+                4 (or "sha1")    = Use plain hex SHA1.
+
+                5 (or "drupal7") = Use Drupal7 salted passwords
+
+                6 (or "joomla15")
+
+                7 (or "ssha")
+
+                8 (or "sha512")
+
+                9 (or "sha256")
+              '';
+            };
+            cryptDefault = mkOption {
+              type = types.nullOr (types.enum [ "md5" "sha256" "sha512" "blowfish" ]);
```

Comment:
I don't think we should allow turning of encryption.

Reply:
As said in the description, this is only applied when `passwordCrypt`is set to 1 or `Y`. So the `null` does not disable the encryption.

Reply:
What is the usecase for null then?

Reply:
The usecase for null is when the user selects another encryption method than 1/Y. For example when the user selects 2/mysql, this option will do nothing.

But a user can select 0/plain if the passwords in the database are unencrypted and I think we should let this option available if for some reasons a old SQL-Schema requires that.

Reply:
The hashing algorithm specified in `cryptDefault` is only used when `crypt` is set to `1` or `Y`. Because in this case, the passwords in the `passwordColumn` (again: this is just a column name) are hashed by the MySQL `CRYPT()` function, which needs the hashing algorithm as a parameter.

If the passwords are hashed for example by using the MySQL `PASSWORD()` function, then we would select `2` or `mysql` as the value for `crypt`.

In this case, hashing algorithm from `cryptDefault` is ignored by `pam_mysql`, because it only needs it when the passwords are hashed with the `CRYPT()` function which only happens when `crypt` is set to `1` or `Y`. So `cryptDefault` is nullable for when `crypt` is set to anything except `1` or `Y`.

So now to the plain passwords: plain passwords are possible in `pam_mysql` and they are used when `crypt` is set to `0` or `plain`.

As already said, I think we can still allow this setting, because maybe someone uses a veeery old SQL schema (for example from a proprietary software) which saves the password in plaintext. Additionally, there is the "HIGHLY DISCOURAGED" in the description of `crypt`.

Reply:
Seems reasonable.